controller/src/main/java/io/pravega/controller/server/ControllerServiceStarter.java

```diff
@@ -253,6 +262,11 @@ protected void shutDown() throws Exception {
                 log.info("Controller cluster listener shutdown");
             }
 
+            if (autoRetentionService != null) {
+                log.info("Awaiting termination of auto retention");
```

Comment:
You are not awaiting here. You only await when you call `awaitTerminated`

Reply:
May I suggest a more compact way of doing this (up to you to do it)? 

You can use `Futures.stopAsync` to asynchronously stop your components, which will give you back a Future, that you can later on use to block. You can also merge multiple such futures into a List and then call `Futures.allOf(...).join()`.

Reply:
this was a typo. we do await terminated later. 

I like the suggestion of using futures.stopAsync. 
But that is an unrelated change. I will open a separate issue for the same and address it.